models completed

devin-w-hill opened the pull request:
SYMCO5. Relationships between models and field names finalized.

Commit:
set changes to pick up on other comp

devin-w-hill committed:
cleaned up most models. error in student.py on foreignKey relationship with researcher.py. do not attempt to make migrations

devin-w-hill committed:
cleaned up most models. error in student.py on foreignKey relationship with researcher.py. do not attempt to make migrations

devin-w-hill committed:
changed Student.researcher field to added_by.

ericharper1 commented on simcon_project/conversation_templates/models/conv_template.py:
```diff
@@ -1,10 +1,12 @@
     name = models.CharField(max_length=40)
     description = models.CharField(max_length=4000)
     creation_date = models.DateTimeField(default=timezone.now)
-    nodes = models.ForeignKey(TemplateNode, related_name='conversation_template')
```
Why did we decide to get rid of this. How will templates point to other templates?

ericharper1 commented on simcon_project/users/admin.py:
```diff
@@ -28,4 +29,11 @@ class CustomUserAdmin(UserAdmin):
+admin.site.register(CustomUser, CustomUserAdmin)
+
+models = apps.get_models()
+for model in models:
```
Can we also do something like this in conversation_templates/admin.py?

devin-w-hill replied:
We created an admin.py in the nested simcon_project so we wouldn't have to do an admin.py for both apps. The way we have it set up in that file seems to work and we actually need to register the user models app there too, we didn't include those for some reason.

ericharper1 replied:
Ahh I see that, that's a good idea

ericharper1 commented on simcon_project/users/models/researcher.py:
```diff
@@ -1,9 +1,7 @@
 
 
 class Researcher(CustomUser):
-    templates = models.ForeignKey(ConversationTemplate, on_delete=models.CASCADE)
```
Why have we gotten rid of all these fields?

devin-w-hill replied:
we found out that the way we had some of the relationships set up was wrong so we moved things around. It ended up that researcher doesn't have any unique fields but we left the class in case we want to add something to it later.

ericharper1 replied:
Okay, that sounds good

ericharper1 commented on simcon_project/users/models/student.py:
```diff
@@ -1,13 +1,9 @@
 
 
 class Student(CustomUser):
-    template_response = models.ForeignKey(TemplateResponse, on_delete=models.CASCADE)
```
I assume this was similar reasoning that was applied to Researcher, but why are we getting rid of all of these fields?

devin-w-hill replied:
we had the relationship backwards for a lot of the fields. ForeignKey is used to create a Many to One relationship and there isnt a One to Many option so that forces us to create the relationship in the model that is the "Many" side. Specifically, we moved the line that creates the relationship between template responses and a student to the template_responses model. We had to do it that way because there are Many responses to one student so the relationship has to be created in that model, not the Student model.

ericharper1 reviewed (commented):
Everything looks good, but I am confused on some changes made

devin-w-hill commented:
Just realized that we need to add the users models to the admin.py in simcon_project and change the related name for the added_by relationship to students from researcher. Will do after this is approved.

devin-w-hill committed:
gave foreignkeys a default of 0. Could not makemigrations without settings defaults and they needed to integers.

devin-w-hill committed:
gave all UUIDfields a default of uuid.uuid4. had to import uuid for all models with UUIDfield

Likhovodov commented on simcon_project/conversation_templates/models/template_node_choice.py:
```diff
@@ -1,7 +1,9 @@
     choice_text = models.CharField(max_length=500)
-    node_id = models.OneToOneField(TemplateNode, related_name='template_node_choice')
\ No newline at end of file
+    destination_node = models.ForeignKey('conversation_templates.TemplateNode', default=0, related_name='destination_choices', on_delete=models.DO_NOTHING)
```
Wondering if related_name should be 'parent_choices' or 'destination_for_choices'.... or maybe something else. Any ideas team? I don't think it will be used either way, but just in case it does end up being used.

ericharper1 replied:
I think 'parent_choices' makes the most sense of those two.

Likhovodov commented on simcon_project/conversation_templates/models/template_node_response.py:
```diff
@@ -1,8 +1,9 @@
+import uuid
 
 
 class TemplateNodeResponse (models.Model):
```
This should have a foreign key to TemplateResponse (called 'parent_template_response' with related name 'node_responses' or something like that). It should cascade on delete (If TemplateResponse goes so should the node responses under that Template response)

Likhovodov commented on simcon_project/conversation_templates/models/template_response.py:
```diff
@@ -1,9 +1,10 @@
\ No newline at end of file
+    id = models.UUIDField(unique=True, editable=False, primary_key=True, default=uuid.uuid4)
+    completion_date = models.DateTimeField(default=None)
+    student = models.ForeignKey('users.Student', related_name='template_responses', default=0, on_delete=models.DO_NOTHING)
```
This should cascade.

Likhovodov commented on simcon_project/users/models/assignment.py:
```diff
@@ -1,13 +1,16 @@
-    templates = models.ForeignKey(ConversationTemplate, related_name='assignments')
-    students = models.ManyToMany(Student, related_name='assignments')
-    labels = models.ManyToMany(SubjectLabel, related_name='assignments')
+    templates = models.ManyToManyField('conversation_templates.ConversationTemplate', related_name='assignments')
```
I think this should be called conversation_templates. Its kind of clunky but I think its worth committing to calling these references by the name of the referred model for consistency.

Same goes for subject_labels in this model.

Likhovodov commented on simcon_project/users/models/subject_label.py:
```diff
@@ -1,10 +1,12 @@
 class SubjectLabel(models.Model):
-    id = models.UUIDField(unique=True, editable=False, primary_key=True)
+    id = models.UUIDField(unique=True, editable=False, primary_key=True, default=uuid.uuid4)
     file_name = models.CharField(max_length=100)
```
label_name would be better.

Likhovodov commented on simcon_project/users/models/subject_label.py:
```diff
@@ -1,10 +1,12 @@
     file_name = models.CharField(max_length=100)
-    students = models.ManyToManyField(Student, related_name='subject_labels')
+    students = models.ManyToManyField('users.Student', related_name='labels')
+    researcher = models.ManyToManyField('users.Researcher', related_name='labels')
```
ForeignKey would be better (labels will not be shared between researchers.

Likhovodov reviewed (commented):
Looking good. There have been a lot of hands on this PR so I hope we get several eyes on this. Thanks for driving the work and tidying things up Devin.

rosasck commented on simcon_project/users/models/student.py:
```diff
@@ -1,13 +1,9 @@
-    def get_is_registered(self):
-        return self.registered
+    added_by = models.ForeignKey('users.Researcher', default=0, related_name='researcher', on_delete=models.CASCADE)
+# needs fixing
```
Are these errors fixed?

rosasck reviewed (changes requested):
Fix changes from David's PR, I would also love to see comments in each file explaining what its used for! Let me know if this is something we want and I can help type them up

devin-w-hill committed:
Changed related name for destination_node

from destination_choices to parent_choices

devin-w-hill committed:
added foreignkey to TemplateResponse

devin-w-hill committed:
updated on_delete behavior

from DO_NOTHING to CASCADE

devin-w-hill committed:
updated field names

from templates to conversation_templates, from labels to subject_labels

devin-w-hill committed:
update field name

from file_name to label_name

devin-w-hill committed:
changed relationship to Researcher

Changed from ManyToManyField to ForeignKey because labels will not be shared between researchers.

devin-w-hill committed:
removed TODO

accessor clash has been resolved so comment stating error was removed.

devin-w-hill reviewed (commented):
Made changes as suggested, pretty sure I got everything. still no comments, working on those now

Likhovodov commented:
> Fix changes from David's PR, I would also love to see comments in each file explaining what its used for! Let me know if this is something we want and I can help type them up

Comments would be good but are not crucial. It would be nice if we could merge this in tonight. If comments are not done, it's not a big deal we can wrap them up later.

devin-w-hill committed:
Added docstring 

Simple docstring describes the class and its fields.

devin-w-hill reviewed (commented):
Updated conv_template.py with a docstring. @rosasck Thoughts?

devin-w-hill commented:
I think it's good to go.

Likhovodov commented:
> I think it's good to go.

I agree.

devin-w-hill commented:
freaking send it bro
Edit: nm we should wait for other people to get a look at it.

Likhovodov commented:
SHIP IT

kaigoos commented:
Looked it over and everything looks good to me!

devin-w-hill commented:
Just FYI everyone, I keep pulling this and having to wipe all my previous migrations before I can runserver without issue. Not sure what I'm doing wrong but it's something to keep in mind if you have issues.
